[lldb][lldb-dap] show modules pane if supported by the adapter

da-viper opened the pull request:
Fixes #140589
Added logic to dynamically set the `lldb-dap.showModules` context based on the presence of modules in the debug session.

da-viper committed:
[lldb][lldb-dap] show modules pane if supported by the adapter

Fixes #140589
Added logic to dynamically set the `lldb-dap.showModules` context based on the presence of modules in the debug session.

JDevlieghere commented on lldb/tools/lldb-dap/src-ts/debug-session-tracker.ts:
```diff
@@ -102,6 +111,8 @@ export class DebugSessionTracker
           break;
       }
+
+      this.showModulesTreeView(modules.length > 0);
```
Does this mean that if all modules get unloaded, the window will now close? I feel like once we've received a module event, we should just keep it open. What do you think?

da-viper replied:
I assumed if there is no modules the session has ended, but changed it since we can have multiple sessions.

Should stay open until there is no more sessions.

JDevlieghere reviewed (commented):
Thanks for the quick turnaround!

da-viper committed:
[lldb][lldb-dap] show modules pane if supported by the adapter

Added logic to dynamically set the `lldb-dap.showModules` context based on the presence of modules in the debug session.

da-viper committed:
[lldb][lldb-dap] show modules pane if supported by the adapter

Added logic to dynamically set the `lldb-dap.showModules` context based on the presence of modules in the debug session.

JDevlieghere commented on lldb/tools/lldb-dap/src-ts/debug-session-tracker.ts:
```diff
@@ -70,9 +71,21 @@ export class DebugSessionTracker
   /** Clear information from the active session. */
   private onExit(session: vscode.DebugSession) {
     this.modules.delete(session);
+    // close when there is no more sessions
```
```suggestion
    // Close when there are no more sessions.
```

JDevlieghere commented on lldb/tools/lldb-dap/src-ts/debug-session-tracker.ts:
```diff
@@ -70,9 +71,21 @@ export class DebugSessionTracker
   private onExit(session: vscode.DebugSession) {
     this.modules.delete(session);
+    // close when there is no more sessions
+    if (this.modules.size <= 0) {
```
Maybe this is common in Javascript and I just don't know, but why not just check `== 0`?

da-viper replied:
It old practice from C

JDevlieghere reviewed (approved):
LGTM modulo comments.

ashgti commented on lldb/tools/lldb-dap/src-ts/debug-session-tracker.ts:
```diff
@@ -70,9 +71,21 @@ export class DebugSessionTracker
   }
 
+  private showModulesTreeView(showModules: boolean) {
+    vscode.commands.executeCommand(
+      "setContext",
+      "lldb-dap.showModules",
+      showModules,
+    );
```
This will set the context globally, however the we should be checking the active debug session and updating this value. Otherwise this UI will show the list for any debug session with modules, not the active one.

We can add a listener for `vscode.debug.onDidChangeActiveDebugSession` to update this value and our module tracker is tracking by session, so we can use the session to track which modules are associated with the active session.

That or the we should move this logic into the `ModulesDataProvider` which is checking the `vscode.debug.activeDebugSession` already.

ashgti commented on lldb/tools/lldb-dap/package.json:
```diff
@@ -787,7 +787,7 @@
           "id": "lldb-dap.modules",
           "name": "Modules",
-          "when": "inDebugMode && debugType == 'lldb-dap'",
+          "when": "inDebugMode && debugType == 'lldb-dap' && lldb-dap.showModules",
```
Can we also add a welcome view for this when its empty and still displayed? 

https://code.visualstudio.com/api/references/contribution-points#contributes.viewsWelcome covers those but a basic welcome message could be:

```json
    "viewsWelcome": [
      {
        "view": "lldb-dap.modules",
        "contents": "No modules are loaded for the active debug session."
      }
    ]
```

It should hide as soon as the tree view has contents.

da-viper replied:
I don't think we should because, each target has at least one module. This applies when launching, attaching and loading a core file.

da-viper committed:
[lldb][lldb-dap] handle multiple debug sessions correctly.

ashgti commented on lldb/tools/lldb-dap/src-ts/debug-session-tracker.ts:
```diff
@@ -35,14 +35,24 @@ export class DebugSessionTracker
+    this.modulesChanged.event;
+
+  constructor() {
+    this.onDidChangeModules(this.moduleChangedListener, this);
```
Its a little odd to have this self referential event emitter. Line 136 could just call `moduleChangedListener` directly.

da-viper replied:
It also responds to the onExit event on line 81
